src/quality-checker.js

```diff
@@ -0,0 +1,116 @@
+const { spawn } = require('child_process');
+const fs = require('fs');
+const readline = require('readline');
+const path = require('path');
+const Papa = require('papaparse');
+
+module.exports = {checkCode, processCouplingMetric};
+
+const papaparseConfig = {
+    header: true,
+    skipEmptyLines: true
+};
+let problemsFound = [];
+
+/**
+ * Uses PMD to check the Java code in the sourceDirectory, finds all the problems from the rule set
+ * and outputs it into the JSON file at outputFileLocation for the front-end to read later,
+ * returns a promise to indicate when it's done or what errors there were
+ * @param {String} sourceDirectory reads code from here
+ * @param {String} outputFileLocation location of the output JSON file
+ * @returns {Promise}
+ */
+function checkCode(sourceDirectory, outputFileLocation) {
+    return new Promise((resolve, reject) => {
+        const pmd = spawn('resources/pmd/bin/run.sh',
+            ['pmd', '-d', sourceDirectory, '-R', 'resources/rulesets/quickstart.xml',
+                '-f', 'csv', '-failOnViolation', 'false']);
+
+        let csvStream = pmd.stdout.pipe(Papa.parse(Papa.NODE_STREAM_INPUT, papaparseConfig));
+        csvStream.on('data', (problem) => {
+            problemsFound.push(problem);
+        });
+
+        pmd.stderr.on('data', (data) => {
+            console.log(`stderr: ${data}`);
+        });
+
+        pmd.on('close', (code) => {
+            console.log(`pmd child process exited with code ${code}`);
+            fs.writeFile(outputFileLocation, JSON.stringify(problemsFound), (error) => {
+                if (error) {
+                    reject(error);
+                }
+            });
+            resolve();
+        });
+    })
+}
+
+/**
+ * Gets the number of times a class in mentioned in another class for all Java files in sourceDirectory
+ * @param {String} sourceDirectory reads code from here
+ */
```

Comment:
props for being so proper l0l

Reply:
lool, i forgot the return value type

shoutouts to vscode/intellij for doing the template for me whenever i type `/**`